src/tokenizer/split_compound_words.rs

```diff
@@ -94,29 +96,31 @@ impl TokenFilter for SplitCompoundWords {
 pub struct SplitCompoundWordsFilter<T> {
     dict: AhoCorasick,
     inner: T,
+    cuts: Vec<usize>,
+    parts: Vec<Token>,
 }
 
 impl<T: Tokenizer> Tokenizer for SplitCompoundWordsFilter<T> {
-    type TokenStream<'a> = SplitCompoundWordsTokenStream<T::TokenStream<'a>>;
+    type TokenStream<'a> = SplitCompoundWordsTokenStream<'a, T::TokenStream<'a>>;
 
     fn token_stream<'a>(&'a mut self, text: &'a str) -> Self::TokenStream<'a> {
         SplitCompoundWordsTokenStream {
             dict: self.dict.clone(),
             tail: self.inner.token_stream(text),
-            cuts: Vec::new(),
-            parts: Vec::new(),
+            cuts: &mut self.cuts,
+            parts: &mut self.parts,
         }
     }
 }
 
-pub struct SplitCompoundWordsTokenStream<T> {
+pub struct SplitCompoundWordsTokenStream<'a, T> {
     dict: AhoCorasick,
     tail: T,
-    cuts: Vec<usize>,
-    parts: Vec<Token>,
+    cuts: &'a mut Vec<usize>,
+    parts: &'a mut Vec<Token>,
 }
 
-impl<T: TokenStream> SplitCompoundWordsTokenStream<T> {
+impl<'a, T: TokenStream> SplitCompoundWordsTokenStream<'a, T> {
     // Will use `self.cuts` to fill `self.parts` if `self.tail.token()`
     // can fully be split into consecutive matches against `self.dict`.
     fn split(&mut self) {
```

Comment:
I think that if the stream is not fully consumed, `parts` could start out non-empty for the next stream? Hence I would suggest calling `self.parts.clear()` next to `self.cuts.clear()` within `split`.

Reply:
Ah right, `advance` already requires `parts` to be empty to actually end up calling `split`. So doing it up front is necessary. 👍🏽